MEN-3334: add docs about static and dynamic deployments

tranchitella opened the pull request.

drewmoseley commented:
LGTM

estenberg commented on 02.Architecture/05.Deployments/docs.md:
```diff
@@ -0,0 +1,81 @@
+    category: docs
+---
+
+To deliver a Mender Artifact to one or more devices using the Mender backend, you have to define a Deployment.
```
Nitpick: "you have to define a Deployment" sounds a bit negative, could rephrase to:

"A Deployment ensures delivery of a Release to one or more devices. A Release can contain one or more Mender Artifacts, one of the Artifacts is assigned when the device receives the update based on software and hardware compatibility."

estenberg commented on 02.Architecture/05.Deployments/docs.md:
```diff
@@ -0,0 +1,81 @@
+
+The Mender UI also supports the creation of a Deployment targeting a static device group, to update all the devices belonging to it.
+
+A **static group**, which is the first type of group that Mender has supported, contains a list of Device IDs. Devices can explicitly be added or removed from a group manually in the UI, or through automation by using the APIs.
```
Would remove " which is the first type of group that Mender has supported," this will soon be irrelevant.

estenberg replied:
Nitpick: I think we usually use italic for defining taxonomy, not bold, for consistency.

estenberg commented on 02.Architecture/05.Deployments/docs.md:
```diff
@@ -0,0 +1,81 @@
+
+Before a device can be added or removed to a static group, it needs to exist in the Mender server. A static deployment is simply a deployment to a static group, so it contains a list of devices and finishes once all the devices in the Deployment have finished.
+
+Please note that Mender uses the group to retrieve the list of the Device IDs at creation time, and devices assigned to the group after the creation of the Deployment won't be included in it.
```
Nitpick: won't -> will not

estenberg commented on 02.Architecture/05.Deployments/docs.md:
```diff
@@ -0,0 +1,81 @@
+
+Mender Enterprise also supports the creation of Deployments using a dynamic filter to assign a deployment to an open, variable group of devices fulfilling the filter criteria as long as the Deployment is not closed.
+
+A **dynamic group** does not contain a list of predefined Device IDs. It instead contains the definition of a dynamic filter that can match one or more devices, and the matched devices could change in the future as new devices are accepted into Mender or device attributes change and fulfill the filter criteria. A dynamic deployment is simply a deployment to a dynamic group. 
```
Italic, not bold?

estenberg replied:
Would perhaps remove "A dynamic deployment is simply a deployment to a dynamic group. ". Perhaps we don't have to define Dynamic deployment?

estenberg commented on 02.Architecture/05.Deployments/docs.md:
```diff
@@ -0,0 +1,81 @@
+
+A **dynamic group** does not contain a list of predefined Device IDs. It instead contains the definition of a dynamic filter that can match one or more devices, and the matched devices could change in the future as new devices are accepted into Mender or device attributes change and fulfill the filter criteria. A dynamic deployment is simply a deployment to a dynamic group. 
+
+Dynamic deployments behave very differently than a static deployment in that it will never implicitly finish, but stay active until the user stops it because we do not necessarily know up-front how many devices are part of this Deployment. It is, however, possible to set a maximum number of devices so that the Deployment finishes after this number of devices have been updated. A device can only b ...
```
Deployments to dynamic groups behave...deployments to a static group..

estenberg replied:
"A device can only be in one static group at the same time, while it can be in multiple dynamic groups." - this is very important information (that we had not documented before).

Perhaps we should rather split this doc into sections like this?
- Static group (what is it, a device can only be in one, etc.)
- Dynamic group (what is it, device can be in multiple, etc.)
- Deployments to static and dynamic groups

estenberg commented on 02.Architecture/05.Deployments/docs.md:
```diff
@@ -0,0 +1,81 @@
+
+Dynamic deployments behave very differently than a static deployment in that it will never implicitly finish, but stay active until the user stops it because we do not necessarily know up-front how many devices are part of this Deployment. It is, however, possible to set a maximum number of devices so that the Deployment finishes after this number of devices have been updated. A device can only b ...
+
+## Lifecycle of a Deployment
```
Deployment lifecycle
(or "The lifecycle of a Deployment")

estenberg commented on 02.Architecture/05.Deployments/docs.md:
```diff
@@ -0,0 +1,81 @@
+
+When all the devices included in the Deployment finished the update, either successfully or not, the Deployment's status changes to "**finished**".
+
+Dynamic deployments behave in a different way as they do not include the explicit list of devices targeted by the update. For this reason, a Dynamic deployment remains in **pending** status until one of the following conditions is true:
```
Deployments to dynamic groups...

mirzak commented on 02.Architecture/05.Deployments/docs.md:
```diff
@@ -0,0 +1,81 @@
+
+To deliver a Mender Artifact to one or more devices using the Mender backend, you have to define a Deployment.
+
+Deployments in Mender represent the explicit association of a release (Mender artifact) and one or more devices running the Mender client.
```
`Mender artifact` -> `Mender Artifact`

mirzak replied:
Nitpick,

`one or more devices running the Mender client.`

Actually it does not have to be **the** Mender client, and instead any client that implements the Device APIs is valid here.

Could be phrased,

```
and one or more devices running a client that implements the Device APIs (link to section in docs), e.g the Mender client
```

tranchitella commented:
@mirzak @estenberg I've updated the text as you suggested. I've split the description of static/dynamic groups from the sections which describe the deployment to static/dynamic groups, it makes the descriptions clearer and easier to understand.

mirzak commented on 02.Architecture/05.Deployments/docs.md:
```diff
@@ -0,0 +1,87 @@
+
+When all the devices included in the Deployment finished the update, either successfully or not, the Deployment's status changes to "*finished*".
+
+Deployments to dynamic groups behave in a different way as they do not include the explicit list of devices targeted by the update. For this reason, a Deployment to dynamic groups remains in *pending* status until one of the following conditions is true:
```
> a Deployment to dynamic groups remains in *pending*

I would expect this to be "inprogress"?

tranchitella replied:
@mirzak nope, they go back to "pending" state if no update is in progress.

mirzak commented on 02.Architecture/05.Deployments/docs.md:
```diff
@@ -0,0 +1,87 @@
+
+When at least one device running the Mender client is performing the defined update, the Deployment transitions to the status "*inprogress*".
+
+When all the devices included in the Deployment finished the update, either successfully or not, the Deployment's status changes to "*finished*".
```
I think we should clarify that this only applies to static deployments
